### PR TITLE
mpl/gpu: Implement optimized IPC cache for ze backend (PR #4) 

### PR DESCRIPTION
**Pull request by abrooks98**
## Pull Request Description

This PR implements an optimized cache mechanism for the ze backend.

Level zero provides a unique id (`mem_id`) for each GPU allocation. As a result, instead of relying on an AVL tree based on pointers, we use this `mem_id` to implement a hashing mechanism for created IPC handles and mapped IPC buffers.

Depends on #6525 

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by abrooks98** [comment]
test:mpich/ch4/gpu

**Comment by abrooks98** [comment]
Ready for review

File: src/mpl/src/gpu/mpl_gpu_ze.c [hunk 1]
```diff
@@ -836,16 +833,19 @@ int MPL_gpu_ipc_handle_create(const void *ptr, MPL_gpu_ipc_mem_handle_t * ipc_ha
             entry->handle = handle;
             HASH_ADD_PTR(gem_hash, ptr, entry, MPL_MEM_OTHER);
         }
+
+        memcpy(&h.fd, &handle, sizeof(int));
```
Comment (raffenet) [inline src/mpl/src/gpu/mpl_gpu_ze.c]: can we just use `h.fd = handle`?

**Review by raffenet — COMMENTED** [review]
a couple of minor comments

File: src/mpl/src/gpu/mpl_gpu_ze.c [hunk 2]
```diff
@@ -793,26 +797,29 @@ int MPL_gpu_finalize(void)
 {
     int i;
 
-    for (i = 0; i < local_ze_device_count; ++i) {
-        MPL_ze_mapped_buffer_entry_t *entry = NULL, *tmp = NULL;
-        if (ipc_cache_removal[i]) {
-            HASH_ITER(hh, ipc_cache_removal[i], entry, tmp) {
-                MPL_gpu_ipc_handle_unmap(entry->ipc_buf);
+    if (likely(MPL_gpu_info.specialized_cache)) {
```
[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpl/src/gpu/mpl_gpu_ze.c:800]: Ah, OK. I see that now. I guess I'm skeptical of `likely` having any impact here, but if it's localized to the ZE implementation its not a big issue.

File: src/mpid/ch4/shm/ipc/gpu/gpu_post.c [hunk 3]
```diff
@@ -8,14 +8,19 @@
...
-        By default, we will cache ipc handle. To manually disable ipc
-        handle cache, user can set this variable to 0.
+      description : |-
+        By default, we will cache ipc handles using the specialized cache mechanism. If the
+        gpu-specific backend does not implement a specialized cache, then we will fallback to
+        the generic cache mechanism. Users can optionally force the generic cache mechanism or
+        disable ipc caching entirely.
+        generic - use the cache mechanism in the generic layer
+        specialized - use the cache mechanism in a gpu-specific mpl layer (if applicable)
+        disabled - disable caching completely
```
Comment (hzhou) [inline src/mpid/ch4/shm/ipc/gpu/gpu_post.c:23]: Wrong commit title ("mpl/gpu: enable mpl-specific cache mechanisms"). I think it more fitting with "ch4/gpu: default MPIR_CVAR_CH4_IPC_GPU_HANDLE_CACHE to specialized"

Comment (abrooks98) [reply src/mpid/ch4/shm/ipc/gpu/gpu_post.c:23]: Fixed.

There will be no impact to non-ze path. Before calling `MPL_gpu_init`, we set the `specialized_cache` member of `MPL_gpu_info` based on the CVAR:

```
        bool specialized_cache =
            (MPIR_CVAR_CH4_IPC_GPU_HANDLE_CACHE == MPIR_CVAR_CH4_IPC_GPU_HANDLE_CACHE_specialized);

        MPL_gpu_info.specialized_cache = specialized_cache;

        int mpl_errno = MPL_gpu_init(debug_summary);
        MPIR_ERR_CHKANDJUMP(mpl_errno != MPL_SUCCESS, mpi_errno, MPI_ERR_OTHER, "**gpu_init");
```

Then check the value after `MPL_gpu_init`. If it is `false` (which is set for fallback, CUDA, and HIP), then it falls back to generic cache (i.e. AVL cache):

```
            /* If the MPL gpu backend doesn't support specialized cache, fallback to generic. */
            if (specialized_cache && !MPL_gpu_info.specialized_cache) {
                MPIR_CVAR_CH4_IPC_GPU_HANDLE_CACHE = MPIR_CVAR_CH4_IPC_GPU_HANDLE_CACHE_generic;
            }
```

File: src/mpl/src/gpu/mpl_gpu_ze.c (hunk 1 above)
Comment (abrooks98) [reply src/mpl/src/gpu/mpl_gpu_ze.c]: Fixed

**Comment by abrooks98** [comment]
test:mpich/warnings

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most
test:mpich/ch4/gpu

**Review by hzhou — APPROVED** [review]
LGTM

**Review by raffenet — APPROVED** [review]

**Pull request merged**